File: sdk/npm/src/providers/chatgpt-org-uk.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";
import { randomInt } from "crypto";

const CHANNEL: Channel = "chatgpt-org-uk";
const BASE_URL = "https://mail.chatgpt.org.uk/api";

const DEFAULT_HEADERS: Record<string, string> = {
  "User-Agent":
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Safari/537.36 Edg/150.0.0.0",
  Accept: "*/*",
  Referer: "https://mail.chatgpt.org.uk/zh/",
  Origin: "https://mail.chatgpt.org.uk",
  DNT: "1",
  "sec-fetch-dest": "empty",
  "sec-fetch-mode": "cors",
  "sec-fetch-site": "same-origin",
};

function randomUsername(length: number = 10): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  let out = "";
  for (let i = 0; i < length; i++) {
    out += chars[randomInt(chars.length)];
  }
  return out;
}

function extractGmSid(response: Response): string {
  const h = response.headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === "function") {
    for (const line of h.getSetCookie()) {
      const match = line.match(/^gm_sid=([^;]+)/);
      if (match) return match[1];
    }
  }
  const setCookie = response.headers.get("set-cookie") || "";
  const match = setCookie.match(/gm_sid=([^;]+)/);
  return match ? match[1] : "";
}

/**
 * 获取可用域名列表
 */
async function fetchDomains(): Promise<string[]> {
  const response = await fetchWithTimeout(`${BASE_URL}/domains/public`, {
    method: "GET",
    headers: DEFAULT_HEADERS,
  });
  if (!response.ok) {
    throw new Error(`chatgpt-org-uk: 获取域名失败 ${response.status}`);
  }
  const data = (await response.json()) as {
    success: boolean;
    data: { domains: Array<{ domain_name: string; is_active: number }> };
  };
  if (!data.success || !Array.isArray(data.data?.domains)) {
    throw new Error("chatgpt-org-uk: 域名响应格式无效");
  }
  return data.data.domains
    .filter((d) => d.is_active === 1)
    .map((d) => d.domain_name);
}

/**
 * 创建收件箱并获取 inbox token + gm_sid cookie
 */
async function createInbox(email: string): Promise<{ token: string; gmSid: string }> {
  const response = await fetchWithTimeout(`${BASE_URL}/inbox-token`, {
    method: "POST",
    headers: {
      ...DEFAULT_HEADERS,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email }),
  });
  if (!response.ok) {
    throw new Error(`chatgpt-org-uk: 创建收件箱失败 ${response.status}`);
  }
  const gmSid = extractGmSid(response);
  const data = (await response.json()) as {
    success: boolean;
    auth?: { token?: string };
  };
  if (!data.success || !data.auth?.token) {
    throw new Error("chatgpt-org-uk: inbox-token 响应缺少 token");
  }
  return { token: data.auth.token, gmSid };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const domains = await fetchDomains();
  if (domains.length === 0) {
    throw new Error("chatgpt-org-uk: 无可用域名");
  }
  const domain = domains[randomInt(domains.length)];
  const username = randomUsername(10);
  const email = `${username}@${domain}`;
  const { token, gmSid } = await createInbox(email);

  return {
    channel: CHANNEL,
    email,
    token: JSON.stringify({ gmSid, inbox: token }),
  };
}

export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!token) {
    throw new Error("chatgpt-org-uk: token 不能为空");
  }

  let gmSid = "";
  let inboxToken = "";
  try {
    const parsed = JSON.parse(token) as { gmSid?: string; inbox?: string };
    gmSid = parsed.gmSid || "";
    inboxToken = parsed.inbox || "";
  } catch {
    inboxToken = token;
  }

  if (!inboxToken) {
    throw new Error("chatgpt-org-uk: inbox token 缺失");
  }

  /* 如果 gmSid 丢失，重新获取 */
  if (!gmSid) {
    const sess = await createInbox(email);
    gmSid = sess.gmSid;
    inboxToken = sess.token;
  }

  const encodedEmail = encodeURIComponent(email);
  const response = await fetchWithTimeout(
    `${BASE_URL}/emails?email=${encodedEmail}`,
    {
      method: "GET",
      headers: {
        ...DEFAULT_HEADERS,
        Cookie: `gm_sid=${gmSid}`,
        "x-inbox-token": inboxToken,
      },
    },
  );

  if (!response.ok) {
    if (response.status === 401 || response.status === 403) {
      const sess = await createInbox(email);
      const retry = await fetchWithTimeout(
        `${BASE_URL}/emails?email=${encodedEmail}`,
        {
          method: "GET",
          headers: {
            ...DEFAULT_HEADERS,
            Cookie: `gm_sid=${sess.gmSid}`,
            "x-inbox-token": sess.token,
          },
        },
      );
      if (!retry.ok) {
        throw new Error(`chatgpt-org-uk: 获取邮件失败 ${retry.status}`);
      }
      const retryData = (await retry.json()) as { success: boolean; data?: { emails?: any[] } };
      if (!retryData.success) return [];
      return (retryData.data?.emails || []).map((raw: any) =>
        normalizeEmail(raw, email),
      );
    }
    throw new Error(`chatgpt-org-uk: 获取邮件失败 ${response.status}`);
  }

  const data = (await response.json()) as { success: boolean; data?: { emails?: any[] } };
  if (!data.success) return [];
  return (data.data?.emails || []).map((raw: any) =>
    normalizeEmail(raw, email),
  );
}

File: sdk/go/provider/chatgpt_org_uk.go
package provider

import (
	"bytes"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
	tls_client "github.com/bogdanfinn/tls-client"
)

const chatgptOrgUkBaseURL = "https://mail.chatgpt.org.uk/api"

// 请求头，参照官网 Web 客户端；Referer 使用 /zh/ 前缀
var chatgptOrgUkHeaders = map[string]string{
	"Accept":         "*/*",
	"Referer":        "https://mail.chatgpt.org.uk/zh/",
	"Origin":         "https://mail.chatgpt.org.uk",
	"DNT":            "1",
	"sec-fetch-dest": "empty",
	"sec-fetch-mode": "cors",
	"sec-fetch-site": "same-origin",
}

// 用户名字符集：小写字母 + 数字
const chatgptOrgUkUsernameChars = "abcdefghijklmnopqrstuvwxyz0123456789"

func setChatgptOrgUkHeaders(req *http.Request) {
	for k, v := range chatgptOrgUkHeaders {
		req.Header.Set(k, v)
	}
	req.Header.Set("User-Agent", GetCurrentUA())
}

// chatgptOrgUkDomainsResponse 对应 GET /api/domains/public 响应
type chatgptOrgUkDomainsResponse struct {
	Success bool `json:"success"`
	Data    struct {
		Domains []struct {
			DomainName string `json:"domain_name"`
			IsActive   int    `json:"is_active"`
		} `json:"domains"`
	} `json:"data"`
}

// chatgptOrgUkInboxTokenResponse 对应 POST /api/inbox-token 响应
type chatgptOrgUkInboxTokenResponse struct {
	Success bool `json:"success"`
	Data    struct {
		Email string `json:"email"`
	} `json:"data"`
	Auth struct {
		Token string `json:"token"`
	} `json:"auth"`
}

// chatgptOrgUkEmailsResponse 对应 GET /api/emails 响应
type chatgptOrgUkEmailsResponse struct {
	Success bool `json:"success"`
	Data    struct {
		Emails []json.RawMessage `json:"emails"`
		Count  int               `json:"count"`
	} `json:"data"`
}

// chatgptOrgUkPackedToken 打包存储 gm_sid 与 inbox JWT
type chatgptOrgUkPackedToken struct {
	GmSid string `json:"gmSid"`
	Inbox string `json:"inbox"`
}

// chatgptOrgUkRandomUsername 本地生成随机用户名
func chatgptOrgUkRandomUsername(length int) (string, error) {
	var b strings.Builder
	n := big.NewInt(int64(len(chatgptOrgUkUsernameChars)))
	for i := 0; i < length; i++ {
		idx, err := rand.Int(rand.Reader, n)
		if err != nil {
			return "", err
		}
		b.WriteByte(chatgptOrgUkUsernameChars[idx.Int64()])
	}
	return b.String(), nil
}

// chatgptOrgUkExtractGmSid 从响应 set-cookie 中提取 gm_sid
func chatgptOrgUkExtractGmSid(resp *http.Response) string {
	for _, cookie := range resp.Cookies() {
		if cookie.Name == "gm_sid" {
			return cookie.Value
		}
	}
	return ""
}

// chatgptOrgUkFetchDomains 获取可用域名列表（过滤 is_active=1）
func chatgptOrgUkFetchDomains(client tls_client.HttpClient) ([]string, error) {
	req, err := http.NewRequest("GET", chatgptOrgUkBaseURL+"/domains/public", nil)
	if err != nil {
		return nil, err
	}
	setChatgptOrgUkHeaders(req)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "chatgpt-org-uk domains"); err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result chatgptOrgUkDomainsResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	if !result.Success {
		return nil, fmt.Errorf("chatgpt-org-uk: 获取域名失败")
	}

	domains := make([]string, 0, len(result.Data.Domains))
	for _, d := range result.Data.Domains {
		if d.IsActive == 1 && d.DomainName != "" {
			domains = append(domains, d.DomainName)
		}
	}
	if len(domains) == 0 {
		return nil, fmt.Errorf("chatgpt-org-uk: 无可用域名")
	}
	return domains, nil
}

// chatgptOrgUkCreateInbox 调用 inbox-token 创建收件箱，返回 inbox JWT 与 gm_sid
func chatgptOrgUkCreateInbox(client tls_client.HttpClient, email string) (inbox string, gmSid string, err error) {
	payload, err := json.Marshal(map[string]string{"email": email})
	if err != nil {
		return "", "", err
	}

	req, err := http.NewRequest("POST", chatgptOrgUkBaseURL+"/inbox-token", bytes.NewReader(payload))
	if err != nil {
		return "", "", err
	}
	setChatgptOrgUkHeaders(req)
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", "", err
	}
	defer resp.Body.Close()

	if err := CheckHTTPStatus(resp, "chatgpt-org-uk inbox-token"); err != nil {
		return "", "", err
	}

	gmSid = chatgptOrgUkExtractGmSid(resp)

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", "", err
	}

	var result chatgptOrgUkInboxTokenResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return "", "", err
	}
	if !result.Success || result.Auth.Token == "" {
		return "", "", fmt.Errorf("chatgpt-org-uk: inbox-token 响应缺少 token")
	}
	return result.Auth.Token, gmSid, nil
}

// chatgptOrgUkParsePackedToken 解析打包的 token
func chatgptOrgUkParsePackedToken(packed string) (gmSid string, inbox string) {
	t := strings.TrimSpace(packed)
	if strings.HasPrefix(t, "{") {
		var p chatgptOrgUkPackedToken
		if json.Unmarshal([]byte(t), &p) == nil && p.Inbox != "" {
			return p.GmSid, p.Inbox
		}
	}
	return "", packed
}

func ChatgptOrgUkGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()

	domains, err := chatgptOrgUkFetchDomains(client)
	if err != nil {
		return nil, err
	}

	idx, err := rand.Int(rand.Reader, big.NewInt(int64(len(domains))))
	if err != nil {
		return nil, err
	}
	domain := domains[idx.Int64()]

	username, err := chatgptOrgUkRandomUsername(10)
	if err != nil {
		return nil, err
	}
	email := username + "@" + domain

	inbox, gmSid, err := chatgptOrgUkCreateInbox(client, email)
	if err != nil {
		return nil, err
	}

	packed, err := json.Marshal(chatgptOrgUkPackedToken{GmSid: gmSid, Inbox: inbox})
	if err != nil {
		return nil, err
	}

	return &CreatedMailbox{Channel: "chatgpt-org-uk", Email: email, Token: string(packed)}, nil
}

func ChatgptOrgUkGetEmails(email string, token string) ([]NormEmail, error) {
	if token == "" {
		return nil, fmt.Errorf("missing inbox token")
	}
	encodedEmail := url.QueryEscape(email)

	gmSid, inbox := chatgptOrgUkParsePackedToken(token)
	if inbox == "" {
		return nil, fmt.Errorf("chatgpt-org-uk: inbox token 缺失")
	}
	client := HTTPClient()

	// gm_sid 丢失时重新创建 session
	if gmSid == "" {
		var err error
		inbox, gmSid, err = chatgptOrgUkCreateInbox(client, email)
		if err != nil {
			return nil, err
		}
	}

	fetchEmails := func(inboxToken string, sid string) ([]NormEmail, error) {
		req, err := http.NewRequest("GET", chatgptOrgUkBaseURL+"/emails?email="+encodedEmail, nil)
		if err != nil {
			return nil, err
		}
		setChatgptOrgUkHeaders(req)
		req.Header.Set("Cookie", fmt.Sprintf("gm_sid=%s", sid))
		req.Header.Set("x-inbox-token", inboxToken)

		resp, err := client.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()

		if err := CheckHTTPStatus(resp, "chatgpt-org-uk get emails"); err != nil {
			return nil, err
		}

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}

		var result chatgptOrgUkEmailsResponse
		if err := json.Unmarshal(body, &result); err != nil {
			return nil, err
		}
		if !result.Success {
			return nil, fmt.Errorf("failed to get emails")
		}

		return NormalizeRawMessages(result.Data.Emails, email)
	}

	emails, err := fetchEmails(inbox, gmSid)
	if err == nil {
		return emails, nil
	}
	// 401/403 时重新创建 session 后重试一次
	if strings.Contains(err.Error(), ": 401") || strings.Contains(err.Error(), ": 403") {
		refreshed, sid, refreshErr := chatgptOrgUkCreateInbox(client, email)
		if refreshErr != nil {
			return nil, err
		}
		return fetchEmails(refreshed, sid)
	}
	return nil, err
}
